File: typescript/src/tools/featureFlags/getDefinition.ts
import type { z } from "zod";
import type { Context, Tool } from "@/tools/types";
import { FeatureFlagGetDefinitionSchema } from "@/schema/tool-inputs";

const schema = FeatureFlagGetDefinitionSchema;

type Params = z.infer<typeof schema>;

export const getDefinitionHandler = async (context: Context, { flagId, flagKey }: Params) => {
	if (!flagId && !flagKey) {
		return {
			content: [
				{
					type: "text",
					text: "Error: Either flagId or flagKey must be provided.",
				},
			],
		};
	}

	const projectId = await context.getProjectId();

	if (flagId) {
		const flagResult = await context.api
			.featureFlags({ projectId })
			.get({ flagId: String(flagId) });
		if (!flagResult.success) {
			throw new Error(`Failed to get feature flag: ${flagResult.error.message}`);
		}
		return {
			content: [{ type: "text", text: JSON.stringify(flagResult.data) }],
		};
	}

	if (flagKey) {
		const flagResult = await context.api
			.featureFlags({ projectId })
			.findByKey({ key: flagKey });
		if (!flagResult.success) {
			throw new Error(`Failed to find feature flag: ${flagResult.error.message}`);
		}
		if (flagResult.data) {
			return {
				content: [{ type: "text", text: JSON.stringify(flagResult.data) }],
			};
		}
		return {
			content: [
				{
					type: "text",
					text: `Error: Flag with key "${flagKey}" not found.`,
				},
			],
		};
	}

	return {
		content: [
			{
				type: "text",
				text: "Error: Could not determine or find the feature flag.",
			},
		],
	};
};

const tool = (): Tool<typeof schema> => ({
	name: "feature-flag-get-definition",
	description: `
        - Use this tool to get the definition of a feature flag. 
        - You can provide either the flagId or the flagKey. 
        - If you provide both, the flagId will be used.
    `,
	schema,
	handler: getDefinitionHandler,
});

export default tool;

File: python/tools/feature_flags/get_definition.py
import json

from api.client import is_error, is_success
from schema.tool_inputs import FeatureFlagGetDefinitionSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def get_feature_flag_definition_handler(context: Context, params: FeatureFlagGetDefinitionSchema) -> ToolResult:
    # Validate that either flagId or flagKey is provided
    if not params.flagId and not params.flagKey:
        raise ValueError("Either flagId or flagKey must be provided")

    project_id = await context.get_project_id()

    # Use flagId if provided (takes precedence)
    if params.flagId:
        flag_result = await context.api.feature_flags(project_id).get(params.flagId)

        if is_error(flag_result):
            raise Exception(f"Failed to get feature flag: {flag_result.error}")

        assert is_success(flag_result)

        return ToolResult(content=json.dumps(flag_result.data.model_dump()))

    # Use flagKey if provided
    if params.flagKey:
        flag_result = await context.api.feature_flags(project_id).find_by_key(params.flagKey)

        if is_error(flag_result):
            raise Exception(f"Failed to find feature flag: {flag_result.error}")

        assert is_success(flag_result)

        if flag_result.data:
            return ToolResult(content=json.dumps(flag_result.data.model_dump()))
        else:
            return ToolResult(content=f'Error: Flag with key "{params.flagKey}" not found.')

    return ToolResult(content="Error: Could not determine or find the feature flag.")


def get_feature_flag_definition_tool() -> Tool[FeatureFlagGetDefinitionSchema]:
    definition = get_tool_definition("feature-flag-get-definition")

    return Tool(
        name="feature-flag-get-definition",
        description=definition["description"],
        schema=FeatureFlagGetDefinitionSchema,
        handler=get_feature_flag_definition_handler,
    )
